File: Frontend/src/services/productService.js
import axiosClient from '../api/axiosClient';

const productService = {
  getAll: () => {
    return axiosClient.get('/products');
  },
  
  getById: (id) => {
    return axiosClient.get(`/products/${id}`);
  },

  // --- HÀM MỚI: TÌM KIẾM & LỌC SẢN PHẨM PHÂN TRANG ---
  searchProducts: (name, minPrice, maxPrice, page = 0, size = 6) => {
    const params = new URLSearchParams();
    if (name) params.set('name', name);       // Khớp với API backend
    if (minPrice) params.set('minPrice', minPrice);
    if (maxPrice) params.set('maxPrice', maxPrice);
    params.set('page', page);
    params.set('size', size);

    return axiosClient.get(`/products/search?${params.toString()}`); //
  }
};

export default productService;

File: backend/src/main/java/website_do_gia_dung/service/ProductService.java
package website_do_gia_dung.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import website_do_gia_dung.entity.Product;
import website_do_gia_dung.repository.ProductRepository;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;

    public List<Product> getAll() {
        return productRepository.findAll();
    }

    // Lấy tất cả có phân trang
    // page bắt đầu từ 0, size = số sản phẩm mỗi trang, sortBy = tên field
    public Page<Product> getAllPaged(int page, int size, String sortBy) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(sortBy).ascending());
        return productRepository.findAll(pageable);
    }

    // Lọc theo giá có phân trang
    public Page<Product> filterByPrice(double minPrice, double maxPrice, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("price").ascending());
        return productRepository.findByPriceBetween(minPrice, maxPrice, pageable);
    }

    // Tìm kiếm tổng hợp: tên + giá + phân trang
    public Page<Product> search(String name, Double minPrice, Double maxPrice, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("name").ascending());
        return productRepository.search(name, minPrice, maxPrice, pageable);
    }

    public Product getById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy sản phẩm id=" + id));
    }

    public List<Product> getByName(String name) {
        return productRepository.findByNameContaining(name);
    }

    public Product create(Product product) {
        return productRepository.save(product);
    }

    public void delete(Long id) {
        productRepository.deleteById(id);
    }
}
